operator_manifest/resolver.py

```diff
@@ -0,0 +1,131 @@
+import json
+import logging
+import os.path
+import subprocess
+
+
+logger = logging.getLogger(__name__)
+
+
+def resolve_image_reference(image_reference, authfile=None):
+    """
+    Resolve the image reference to a digest image reference.
+
+    :param str image_reference: the image reference of the container image to resolve
+    :param str authfile: the path of the authentication file for registry communication.
+    :return: the image reference resolved to a digest
+    :rtype: str
+    """
+    extra_args = []
+    if authfile:
+        if not os.path.exists(authfile):
+            raise ValueError(f'Specified authfile {authfile} does not exist')
+        extra_args.append('--authfile')
+        extra_args.append(authfile)
+
+    logger.debug('Resolving %s', image_reference)
+    name = _get_container_image_name(image_reference)
+    skopeo_raw = _skopeo_inspect(f'docker://{image_reference}', '--raw', *extra_args)
+    if json.loads(skopeo_raw).get('schemaVersion') == 2:
+        raw_digest = hashlib.sha256(skopeo_raw.encode('utf-8')).hexdigest()
+        digest = f'sha256:{raw_digest}'
+    else:
+        # Schema 1 is not a stable format. The contents of the manifest may change slightly
+        # between requests causing a different digest to be computed. Instead, let's leverage
+        # skopeo's own logic for determining the digest in this case. In the future, we
+        # may want to use skopeo in all cases, but this will have significant performance
+        # issues until https://github.com/containers/skopeo/issues/785
+        digest = json.loads(_skopeo_inspect(f'docker://{image_reference}', *extra_args))['Digest']
+    resolved_image_reference = f'{name}@{digest}'
+    logger.debug('%s resolved to %s', image_reference, resolve_image_reference)
+    return resolved_image_reference
+
+
+def _get_container_image_name(image_reference):
+    """
+    Get the container image name from an image reference.
+
+    :param str image_reference: the image reference to analyze
+    :return: the container image name
+    """
+    if '@' in image_reference:
+        return image_reference.split('@', 1)[0]
+    else:
+        return image_reference.rsplit(':', 1)[0]
+
+
+def _retry(attempts=3, wait_on=Exception):
+    """
+    Retry a section of code until success or max attempts are reached.
+
+    :param int attempts: the total number of attempts to make before erroring out
+    :param Exception wait_on: the exception on encountering which the function will be retried
+    :raises Exception: if the maximum attempts are reached
+    """
+
+    def wrapper(function):
+        @functools.wraps(function)
+        def inner(*args, **kwargs):
+            remaining_attempts = attempts
+            while True:
+                try:
+                    return function(*args, **kwargs)
+                except wait_on:
+                    remaining_attempts -= 1
+                    if remaining_attempts <= 0:
+                        raise
```

Comment:
I was a bit surprised by the fact that there's no delay, I don't think there needs to be one for the current use case but might be worth calling it out in the docstring

Reply:
Specified this in the docstring. The reason I don't think it's needed now is because skopeo already has built-in retry logic. So this is just additional "resiliency".